sysvabi64/sysvabi64.rst

```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+resolver function.
+
+The static relocations with a prefix of ``R_AARCH64_TLSDESC_``
+targeting TLS symbol ``var``, instruct the static linker to create a
+TLS Descriptor for ``var``. The TLS Descriptor for a variable is
+stored in a pair of consecutive GOT entries, N and N + 1. The GOT
+entry for N has a dynamic ``R_AARCH64_TLSDESC`` relocation targeting
+the TLS symbol for ``var``.
+
+When resolving the ``R_AARCH64_TLSDESC`` relocation, the dynamic
+loader places the address of the chosen resolver function in the first
+GOT entry, and the argument for the chosen resolver function in the
+second GOT entry.
+
+The AArch64 C and assembler examples are adapted from the AArch32
+TLSDESC_ paper. The C code below represents the TLS Descriptor.
+
+.. code-block:: c
+
+    // Argument passed to TLS resolver functions.
+    struct tlsdesc
+    {
+      ptrdiff_t (*resolver)(struct tlsdesc *);
+      union
+      {
+        void *pointer;
+        long value;
+      } argument;
+    };
+
+TLS Resolver Functions
+----------------------
+
+The TLS resolver functions are not standardized by this ABI as they
+are internal to the dynamic linker. Programs must not directly refer
+to TLS resolver functions.
+
+Calling Convention
+^^^^^^^^^^^^^^^^^^
+
+TLS resolver functions have one argument, the address of the TLS
+descriptor, passed in ``x0``, they return the offset of the variable
+from the thread pointer in ``x0``.
+
+TLS resolver functions must save all registers that they modify with
+the exception of ``x0``, ``x1``, ``x30`` and the processor flags.
+
+Example Resolver Functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+These examples are for illustrative purposes only. Due to the
+restrictions on calling convention, the resolver routines must be
+written in assembly language.
+
+Static TLS Specialization
+
+When the TLS variable is in the static TLS block, the offset from the
+thread pointer is fixed at runtime. The dynamic loader can calculate
+the offset and place it in the TLS descriptor. All the static TLS
+resolver function needs to do is extract the offset and return it.
+
+.. code-block:: asm
+
+    _dl_tlsdesc_return:
+    // x0 contains pointer to struct tlsdesc.
+    // tlsdesc.argument.value contains offset of variable from TP
+      ldr   x0, [x0, #8]
+      ret
+
+Dynamic TLS Specialization
+
+When the TLS variable is defined in dynamic TLS the address of the TLS
+variable must be calculated by the resolver function using
+``__tls_get_addr``. The resolver function returns the offset from the
+thread pointer by subtracting the address of the thread pointer from
+the address of the TLS variable. In practice an implementation of the
+dynamic TLS resolver contains many platform specific details outside
+of the scope of the ABI. An example of how a dynamic resolver might be
+implemented can be found in the Dynamic Specialization section of
+TLSDESC_.
+
+Undefined Weak Symbols
+
+An undefined weak symbol has the value 0. As the resolver function
```

Comment:
This might be the glibc behavior, but musl doesn't have the special `__dl_tlsdesc_undefweak`. I think it's better to allow flexibility and require a particular behavior on undefined weak TLS.

Reply:
This part is just an example of what can be done. I've written at the top of the section
> The TLS resolver functions are not standardized by this ABI as they are internal to the dynamic linker

and

> These examples are for illustrative purposes only

I'll see if there's anything I can do to state that there is no requirement to implement a specific resolver function.